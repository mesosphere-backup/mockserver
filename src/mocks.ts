import { resolve } from "path";
import { sync as syncGlob } from "glob";
import { IMock } from "./types";
import { Request, RequestHandler, Response } from "express";

function validateMock(file: string, mocks: any): IMock[] {
  if (!(mocks instanceof Array)) {
    throw new Error(`You missed to export mocks in ${file}, please provide at least this default case:
export const mocks = [];`);
  }

  mocks.forEach((mock: any, index: number) => {
    if (!mock.id || typeof mock.id !== "string") {
      throw new Error(
        `The mock export at index ${index} at ${file} does not include a valid id: ${JSON.stringify(
          mock
        )}`
      );
    }

    if (!mock.request || typeof mock.request !== "function") {
      throw new Error(
        `The mock export at index ${index} at ${file} does not include a valid request: ${JSON.stringify(
          mock
        )}`
      );
    }
  });

  return mocks as IMock[];
}

export function getMockForJSON(json: string | {}): RequestHandler {
  if (!json) {
    throw new Error("JSON argument is missing!");
  }

  const jsonString = JSON.stringify(
    typeof json === "string" ? JSON.parse(json) : json
  );

  return (_req: Request, res: Response) => {
    res.setHeader("Content-Type", "Application/json");
    res.send(jsonString);
  };
}

export function discoverMocks(globString: string): IMock[] {
  // @ts-ignore
  // tslint:disable-next-line:no-empty
  global.describe = () => {};

  return syncGlob(globString)
    .map(file => {
      try {
        return validateMock(file, require(resolve(process.cwd(), file)).mocks);
      } catch (e) {
        return [];
      }
    })
    .reduce((carry, item) => carry.concat(item), []);
}
