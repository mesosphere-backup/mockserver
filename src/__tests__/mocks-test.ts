import { discoverMocks, getMockForJSON } from "../mocks";
import { sync } from "glob";
import { resolve } from "path";

jest.mock("glob");
jest.mock("path");

describe("mocks", () => {
  describe("#extractMocks", () => {
    it("fails malformed mock data (missing id)", () => {
      const mockfilename = "1-invalid-mock-without-id";
      sync.mockReturnValue([mockfilename]);
      resolve.mockReturnValue(mockfilename);
      expect(() => {
        discoverMocks(mockfilename);
      }).toThrowErrorMatchingSnapshot();
    });

    it("fails with object given an invalid id", () => {
      const mockfilename = "1-invalid-mock-with-invalid-id";
      sync.mockReturnValue([mockfilename]);
      resolve.mockReturnValue(mockfilename);
      expect(() => {
        discoverMocks(mockfilename);
      }).toThrowErrorMatchingSnapshot();
    });

    it("fails without given mock function", () => {
      const mockfilename = "1-invalid-mock-without-request";
      sync.mockReturnValue([mockfilename]);
      resolve.mockReturnValue(mockfilename);
      expect(() => {
        discoverMocks(mockfilename);
      }).toThrowErrorMatchingSnapshot();
    });

    it("fails with something not a function given as mock", () => {
      const mockfilename = "1-invalid-mock-invalid-request";
      sync.mockReturnValue([mockfilename]);
      resolve.mockReturnValue(mockfilename);
      expect(() => {
        discoverMocks(mockfilename);
      }).toThrowErrorMatchingSnapshot();
    });

    it("fails with mocks not being an array", () => {
      const mockfilename = "1-invalid-mock-not-an-array";
      sync.mockReturnValue([mockfilename]);
      resolve.mockReturnValue(mockfilename);
      expect(() => {
        discoverMocks(mockfilename);
      }).toThrow();
    });

    it("doesnt fail with string given as id", () => {
      const mockfilename = "1-valid-mock";
      sync.mockReturnValue([mockfilename]);
      resolve.mockReturnValue(mockfilename);
      expect(() => {
        discoverMocks(mockfilename);
      }).not.toThrow();
    });

    it("doesnt fail with with empty mocks array", () => {
      const mockfilename = "1-valid-mock-empty-array";
      sync.mockReturnValue([mockfilename]);
      resolve.mockReturnValue(mockfilename);
      expect(() => {
        discoverMocks(mockfilename);
      }).not.toThrow();
    });
  });

  describe("#getMockForJSON", () => {
    it("throws unless json is provided", () => {
      expect(() => {
        getMockForJSON(null);
      }).toThrow();
    });

    it("throws unless valid json string is provided", () => {
      expect(() => {
        getMockForJSON("foo");
      }).toThrow();
    });

    it("doesn't throw when json is provided", () => {
      expect(() => {
        getMockForJSON({});
      }).not.toThrow();
    });

    it("doesn't throw when valid json string is provided", () => {
      expect(() => {
        getMockForJSON("{}");
      }).not.toThrow();
    });

    it("returns doesn't throw when json is provided", () => {
      const mockfilename = "1-valid-json-mock";

      sync.mockReturnValue([mockfilename]);
      resolve.mockReturnValue(mockfilename);

      expect(() => {
        discoverMocks(mockfilename);
      }).not.toThrow();
    });
  });
});
