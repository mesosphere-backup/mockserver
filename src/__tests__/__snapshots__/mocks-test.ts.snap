// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mocks #extractMocks fails malformed mock data (missing id) 1`] = `"The mock export at index 0 at 1-invalid-mock-without-id does not include a valid id: {}"`;

exports[`mocks #extractMocks fails with object given an invalid id 1`] = `"The mock export at index 0 at 1-invalid-mock-with-invalid-id does not include a valid id: {\\"id\\":42}"`;

exports[`mocks #extractMocks fails with something not a function given as mock 1`] = `"The mock export at index 0 at 1-invalid-mock-invalid-request does not include a valid request: {\\"id\\":\\"42\\",\\"request\\":23}"`;

exports[`mocks #extractMocks fails without given mock function 1`] = `"The mock export at index 0 at 1-invalid-mock-without-request does not include a valid request: {\\"id\\":\\"32\\"}"`;
